File: execpad/src/run-log.ts
import type { RunLogEntry } from "./types.js";

/** Serialize session log for tools / eval artifacts. */
export function exportRunLogJSON(entries: readonly RunLogEntry[], pretty = true): string {
  return JSON.stringify([...entries], null, pretty ? 2 : undefined);
}

/** Human-readable log for pasting into issues or LLM context. */
export function exportRunLogMarkdown(entries: readonly RunLogEntry[]): string {
  const lines: string[] = ["# execpad session log", ""];
  for (const e of entries) {
    lines.push(`## ${e.id} — ${e.language} @ ${new Date(e.at).toISOString()}`);
    lines.push("");
    lines.push(`- **exit:** ${e.exitCode} · **duration:** ${e.durationMs}ms · **truncated:** ${e.truncated}`);
    lines.push(`- **cwd:** \`${e.cwd}\``);
    if (e.files.length) {
      lines.push(`- **files:** ${e.files.map((f) => `${f.type}:${f.path}`).join(", ")}`);
    }
    lines.push("");
    lines.push("### code");
    lines.push("");
    lines.push("```");
    lines.push(e.code);
    lines.push("```");
    lines.push("");
    lines.push("### stdout");
    lines.push("");
    lines.push("```");
    lines.push(e.stdout || "(empty)");
    lines.push("```");
    lines.push("");
    lines.push("### stderr");
    lines.push("");
    lines.push("```");
    lines.push(e.stderr || "(empty)");
    lines.push("```");
    lines.push("");
  }
  return lines.join("\n");
}

File: execpad/python/src/agentpad/run_log.py
from __future__ import annotations

import json
from datetime import datetime, timezone

from agentpad.types import RunLogEntry


def export_run_log_json(entries: list[RunLogEntry], *, pretty: bool = True) -> str:
    payload = [
        {
            "id": e.id,
            "at": e.at,
            "language": e.language,
            "code": e.code,
            "cwd": e.cwd,
            "exitCode": e.exit_code,
            "durationMs": e.duration_ms,
            "stdout": e.stdout,
            "stderr": e.stderr,
            "truncated": e.truncated,
            "files": [{"path": f.path, "type": f.type, "size": f.size} for f in e.files],
        }
        for e in entries
    ]
    return json.dumps(payload, indent=2 if pretty else None)


def export_run_log_markdown(entries: list[RunLogEntry]) -> str:
    lines: list[str] = ["# agentpad session log", ""]
    for e in entries:
        ts = datetime.fromtimestamp(e.at / 1000, tz=timezone.utc).isoformat()
        lines.append(f"## {e.id} — {e.language} @ {ts}")
        lines.append("")
        lines.append(
            f"- **exit:** {e.exit_code} · **duration:** {e.duration_ms}ms · **truncated:** {e.truncated}"
        )
        lines.append(f"- **cwd:** `{e.cwd}`")
        if e.files:
            parts = [f"{f.type}:{f.path}" for f in e.files]
            lines.append(f"- **files:** {', '.join(parts)}")
        lines.extend(["", "### code", "", "```", e.code, "```", "", "### stdout", "", "```", e.stdout or "(empty)", "```"])
        lines.extend(["", "### stderr", "", "```", e.stderr or "(empty)", "```", ""])
    return "\n".join(lines)
